pkg/addonfactory/addonfactory.go

```diff
@@ -115,6 +116,13 @@ func (f *AgentAddonFactory) WithConfigGVRs(gvrs ...schema.GroupVersionResource)
 	return f
 }
 
+// WithHostingCluster defines the hosting cluster used in hosted mode. An AgentAddon may use this to provide
+// additional metadata.
+func (f *AgentAddonFactory) WithHostingCluster(cluster *clusterv1.ManagedCluster) *AgentAddonFactory {
+	f.hostingCluster = cluster
```

Comment:
I think in sd case, there could be multiple hosting clusters? In the Manifest func, you could get the hosting cluster for the particular addon and then get that hosting cluster.

Reply:
@qiujian16 it seems that it is expected to have an `AgentAddonFactory` instance per addon based on the `addonName` parameter being required in `NewAgentAddonFactory`.

Since an addon can only be hosted on one hosting cluster, I think the approach in my PR is okay.

Ideally, I would have wanted to populate this automatically based on if the addon specified a hosting cluster, however, `AgentAddonFactory` doesn't have a Kubernetes client and I didn't want to introduce a breaking change of requiring it.

Reply:
I think you could have one hub (where hub addon manager run), and multiple hosting clusters (where many hosted addon agents run on each hosting cluster and they all connect to the same hub)? If it is not the case, I am fine with this fix.

Reply:
@qiujian16 that's possible, but I'm saying there is a 1x1 mapping of `AgentAddonFactory` to addon, so  `AgentAddonFactory`  needs to only be aware of one hosting cluster.

Reply:
@zhiweiyin318 do we have this requirement? do we have doc this?

Reply:
@qiujian16 I say it's a 1x1 mapping because of the function signature to create the `AgentAddonFactory` requires the addon name:
https://github.com/open-cluster-management-io/addon-framework/blob/259bd501bbfb2f22256073eb1c4e2fbf3bc754f6/pkg/addonfactory/addonfactory.go#L41

Reply:
but addon name is per addon on hub I think. if you have multiple hosting clusters and one hub, you could have one addonfactory with a single addon name I think? Or maybe I miss something here.

Reply:
@qiujian16 oh, I see what you are saying. Yes, you are right because the ManagedClusterAddOn object is passed to the `Manifests` method and not to the addon factory.

With that in mind, I'm not quite sure how we could provide this change without changing the signature of an public API.

What do you suggest?

Reply:
I think for now in the Manifests() func of helm agent addon factory, we could build a cluster client, get the hosting cluster and its version. Or even better, we could provide a `ManifestRenderingHook` on the factory, so the addon can read the hosting cluster in the hook func and render the manifest. It is not a perfect solution but I think it is better to defer the signature change until we have more use cases. 

I am thinking a general `ManifestOptions` as the input of Manifest() func, but I cannot find other use case in my to mind to use this option except hosting clusters (and even for hosting cluster, not all addon needs to know the version of hosting cluster). This is reason I feel we should not change signature for now.